main/src/ca/uwaterloo/flix/language/phase/ClosureConv.scala

```diff
@@ -398,4 +433,224 @@ object ClosureConv {
     visitExp(e0)
   }
 
+  /**
+    * Performs closure conversion on a [[Expr.LocalDef]].
+    * Adds any captured variables to the list of formal parameters and rewrites any
+    * [[Expr.ApplyLocalDef]] to include the captured variables or, in the case of a recursive call,
+    * the list of new formal parameters.
+    *
+    * E.g.,
+    *
+    * {{{
+    *   def f(): Int32 = {
+    *       let a = 1;
+    *       let b = 2;
+    *       let c = 3;
+    *       def g(l) = if (l == 4) a + b + c + l else g(4);
+    *       g(4)
+    *   }
+    * }}}
+    *
+    * becomes
+    *
+    * {{{
+    *   def f(): Int32 = {
+    *       let a = 1;
+    *       let b = 2;
+    *       let c = 3;
+    *       def g(x, y, z, l) = if (l == 4) x + y + z + l else g(x, y, z, 4);
+    *       g(a, b, c, 10)
```

Comment:
I think tail recursive functions breaks this: Something like:

```
def foo(x: Int32, r: Unit -> Int32): Int32 -> Int32 = 
  def local() = x;
  if (x == 0) r else foo(x - 1, _ -> local())
 ```
which gets rewritten to a loop where the value of `x` changes (by mutation).

The point I am trying get across is that I will modify `x` between the time when the closure should have been created and when it is currently created.

Reply:
Im pretty sure your example doesn't break because the tail call is the last instruction, so the `_ -> local()` is still a closure creation that happens before the tail call.

Reply:
We can add it as a test

Reply:
Also, it just adds `x` to the list of arguments in the call but, crucially, before the lambda is closure converted so when visiting the lambda `_ -> local()` it will have `x` as a free variable of the body of the lambda and introduce it as a formal parameter of the lambda. That's why we do rewriting of `ApplyLocalDef` before anything else

Reply:
but i think we have good material for Monday 👨‍🔬

Reply:
My counter-example may not be a counter-example, but you still need to convince yourself and me that the transformation is correct (and that it does not rely on informal invariants).

Reply:
not relying on informal invariants seems like a higher constraint then many parts of the compiler 😄